src/libyang.h

```diff
@@ -131,6 +131,22 @@ extern "C" {
  * `next` pointer of the last item in the list is always NULL.
  */
 
+/**
+ * @page howtoThreads Threading Limitations
+ *
+ * @section context Context
+ *
+ * It is safe to read from ::ly_ctx structure concurrently and use its dictionary, which is protected by a lock.
+ * Thread-safe functions include any ones working with data trees and all the `ly_ctx_get_*()` functions. Generally,
+ * they are the functions with `const` context parameter.
+ *
+ * @section data Data Trees
+ *
+ * The general rule is always the same. Only the functions with non-const ::lyd_node parameters modify the node(s).
+ * No concurrent execution of such functions should be allowed on a single data tree or subtrees of one. Conforming
+ * to this, it should be safe to do everything else.
+ */
```

Comment:
Disclaimer: I've never used libyang2 before, but I have some experience with the `1.x` branch of libyang. I think the intention is that there can only ever be one writer or N readers per each tree, but the way it's written may allow interpreting that as "no concurrent writers" while allowing a mixture of R+W workloads.

Reply:
There are no theoretical thread-safety differences between *libyang* 1 and 2. However, there are several use-cases of *libyang* 1 not being thread-safe despite a reasonable expectation for it to be. That should be fixed in *libyang* 2.